### DISCUSSION TITLE
Allow Discogs release ID in "as Tracks"

### DISCUSSION THREAD

**Discussion by martimpassos** [beetbox/beets#4320] [Ideas]
Discogs is a far better metadata source for my collection (old school electronic) than MB, and I'm assuming there might be other niches that share this thought. 

When I encounter a folder with mixed releases I have two options: `as Tracks` and `Group albums`. The former doesn't allow me to use Discogs release IDs, which limits my options to MB. The latter enters a loophole I cannot exit: first it correctly separates the tracks but none of the suggestions are the correct one. As soon as I try to enter an ID for the first two, it brings back the third that belongs to a different release. See below: 

```
Enter search, enter Id, aBort, eDit, edit Candidates? g

/Users/martimpassos/Music/musica/[none] Blackwood - Ride On The Rhythm (2017)/Blackwood - Ride On The Rhythm (90's Remix).mp3
/Users/martimpassos/Music/musica/[none] Blackwood - Ride On The Rhythm (2017)/Blackwood - Ride On The Rhythm (Alex Natale Remix).mp3
/Users/martimpassos/Music/musica/[none] Blackwood - Ride On The Rhythm (2017)/Blackwood - Ride On The Rhythm (Stefano Noferini Remix).mp3 (3 items)
Tagging:
    Blackwood - Ride On The Rhythm
URL:
    https://www.discogs.com/release/1595356-Blackwood-Ride-On-The-Rhythm
(Similarity: 86.5%) (year, tracks, unmatched tracks) (Discogs, Vinyl, 1997, France, Hot Tracks (3), SHT 3356-6)
 * Blackwood - Ride On The Rhythm (Alex Natale Remix) (#0) -> Ride On The Rhythm (Radio Version) (#1) (title)
 * Ride On The Rhythm (90's Remix) (#0)                    -> Ride On The Rhythm (Marascia Remix) (#2) (title)
Unmatched tracks (1):
 ! Blackwood - Ride On The Rhythm (Stefano Noferini Remix) (# 0) (4:55)
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates? i
Enter release ID: [r13678441]
Tagging:
    Blackwood - Ride On The Rhythm
URL:
    https://www.discogs.com/release/13678441-Blackwood-Ride-On-The-Rhythm
(Similarity: 90.0%) (missing tracks, tracks) (Discogs, File, 2017, Genius Groove)
 * Blackwood - Ride On The Rhythm (Alex Natale Remix) (#0)      -> Ride On The Rhythm (Alex Natale Remix) (#1) (title)
 * Ride On The Rhythm (90's Remix) (#0)                         -> Ride On The Rhythm (90's Remix) (#3)
 * Blackwood - Ride On The Rhythm (Stefano Noferini Remix) (#0) -> Ride On The Rhythm (Remix) (#4) (title)
Missing tracks (1/4 - 25.0%):
 ! Ride On The Rhythm (Club Version) (# 2) (6:20)
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates? 
```

**Comment by sampsyo** [comment]
Hmm, I'm not quite sure what we'd do with a release ID when tagging an individual track... are there track IDs in Discogs?

**Reply by martimpassos** [reply]
There aren't. My idea was to tag the track as an incomplete album instead (considering that **all** tracks are contained in releases, this extrapolation sounds reasonable to me). Of course for VAs it would still lack track artist, but a lot of fields could be populated this way...

**Reply by martimpassos** [reply]
In the above case, I could use the same release ID for the first two tracks, then another one for the last track. Assuming that the autotagger manages to identify the track numbers, I'd end up with two correctly tagged partial releases

**Reply by sampsyo** [reply]
Hmm… that seems to make implementation here pretty tricky. If we're trying to match an individual item against an entire album, then we'd need machinery to pick out the best track among all the tracks on the album. That is conceptually possible, but it would mean treating an ID more like another way into a fuzzy search rather than as an "exact match," which is how we usually treat manually-entered IDs. So while I don't want to rule it out entirely, I just want to register a warning here that the implementation might get quite messy…

**Reply by martimpassos** [reply]
I see. I'm not super familiar with how beets works under the hood, but since we expect the user to enter a release that actually contains that track (and that it could be re-entered indefinitely in case of mistakes), I think the human-provided ID could still be used as an "exact match" and plain string similarity could work well in many [cases](https://github.com/beetbox/beets/discussions/4318). If no good match is found, asking the user to indicate the correct track(s) index in the release's full tracklist isn't terrible UX IMO.

Here's a very early UI sketch. In this case `Alex Natale` and `90's` remixes belong to the same release (the one correctly suggested by beets), but `Stefano Noferini` belongs to a different one, hence the need to import them as tracks. Otherwise I'd need to Skip, move `Stefano Noferini` to a different folder, and reimport/tag. Tagging them individually would let `beets move` do this for me later.

```
Tagging:
    Blackwood - Ride On The Rhythm
URL:
    https://www.discogs.com/release/13678441-Blackwood-Ride-On-The-Rhythm
(Similarity: 90.0%) (missing tracks, tracks) (Discogs, File, 2017, Genius Groove)
 * Blackwood - Ride On The Rhythm (Alex Natale Remix) (#0)      -> Ride On The Rhythm (Alex Natale Remix) (#1) (title)
 * Ride On The Rhythm (90's Remix) (#0)                         -> Ride On The Rhythm (90's Remix) (#3)
 * Blackwood - Ride On The Rhythm (Stefano Noferini Remix) (#0) -> Ride On The Rhythm (Remix) (#4) (title)
Missing tracks (1/4 - 25.0%):
 ! Ride On The Rhythm (Club Version) (# 2) (6:20)
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates? t
 * Blackwood - Ride On The Rhythm (Alex Natale Remix)
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates? i
Enter release ID: [r13678441]
* (similarity 90%)Blackwood - Ride On The Rhythm (Alex Natale Remix) (#0)       -> Ride On The Rhythm (Alex Natale Remix) (#1) (title)
*                                                                                  Ride On The Rhythm (Club Mix) (#2)
*                                                                                  Ride On The Rhythm (90's Mix) (#3)
*                                                                                  Ride On The Rhythm (Remix) (#4)
[A]pply, Skip, Provide #, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort? P # say there's no good match or the user knows the file is actually another track
Enter track number: 2
*                                                                                  Ride On The Rhythm (Alex Natale Remix) (#1)
* (similarity 70%)Blackwood - Ride On The Rhythm (Alex Natale Remix) (#0)       -> Ride On The Rhythm (Club Mix) (#2) (title)
*                                                                                  Ride On The Rhythm (90's Mix) (#3)
*                                                                                  Ride On The Rhythm (Remix) (#4)
[A]pply, Skip, Provide #, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort? A
... # by default stay in the same release for the next track ("90's Remix", omitted here), allowing the user to change releases via enter Id:
*                                                                                  Ride On The Rhythm (Alex Natale Remix) (#1)
* (similarity 60%)Blackwood - Ride On The Rhythm (Stefano Noferini Remix) (#0)  -> Ride On The Rhythm (Club Mix) (#2)
*                                                                                  Ride On The Rhythm (90's Mix) (#3)
*                                                                                  Ride On The Rhythm (Remix) (#4)
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates? i
Enter release ID: [r390832] # different Discogs ID
*                                                                                  Ride On The Rhythm (Radio Version Remix By Alex Natale) (#1)
*                                                                                  Ride On The Rhythm (Club Version Remix By Alex Natale) (#2)
* (similarity 88%)Blackwood - Ride On The Rhythm (Stefano Noferini Remix) (#0)  -> Ride On The Rhythm (Remix By Stefano Noferini) (#3) (title)
*                                                                                  Ride On The Rhythm (Remix By Marascia) (#4)
[A]pply, Skip, Provide #, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort? A
```

**Reply by JOJ0** [reply]
> Hmm… that seems to make implementation here pretty tricky. If we're trying to match an individual item against an entire album, then we'd need machinery to pick out the best track among all the tracks on the album. That is conceptually possible, but it would mean treating an ID more like another way into a fuzzy search rather than as an "exact match," which is how we usually treat manually-entered IDs. So while I don't want to rule it out entirely, I just want to register a warning here that the implementation might get quite messy…

Ok, if Tracks mode expects an exact match, do you think it's possible that a discogs id could be passed here, but suffixed with a "dash number" to explicitely tell which track on the Discogs release to map. The suffix should be mandatory. To keep implementation easy, also mandatory should be the format: numerical, no A1, B2 record track numbers allowed!

It's not the most beautiful way but IMO would fix the garantueed dead end when trying to do Tracks mode with Discogs releases. E.g

```
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates? i
Enter recording ID: 390832-2
```

What do you think about this idea @sampsyo? Implementation possible without too much hassle?

**Reply by sampsyo** [reply]
Sure! Inventing some kind of "fake" track ID if Discogs doesn't have them, such as `<album id>-<index>`, could work. It would require some exploratory hacking to see how feasible it is.

**Reply by JOJ0** [reply]
Just a reminder to myself or anybody picking this up: "Fake" track ID's are "a thing" in discogs.py already: https://github.com/beetbox/beets/blob/master/beetsplug/discogs.py#L353-L355
